cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -775,30 +757,31 @@ private String getIndexColumn(String key, SortInfo.SortOrder sortOrder) {
   /**
    * Creates a {@link Put} for the a metadata index
    *
-   * @param targetId the {@link NamespacedEntityId} from which the metadata index has to be created
+   * @param metadataEntity the {@link MetadataEntity} from which the metadata index has to be created
    * @param metadataKey the key of the metadata entry
    * @param index the index for this metadata
    * @param indexColumn the column to store the index in. This column should exist in the
    * {@link IndexedTable#INDEX_COLUMNS_CONF_KEY} property in the dataset's definition
    * @return {@link Put} which is a index row with the value to be indexed in the #indexColumn
    */
-  private Put getIndexPut(NamespacedEntityId targetId, String metadataKey, String index, String indexColumn) {
-    MDSKey mdsIndexKey = MdsKey.getMDSIndexKey(targetId, metadataKey, index.toLowerCase());
-    String namespacedIndex = targetId.getNamespace() + KEYVALUE_SEPARATOR + index.toLowerCase();
+  private Put getIndexPut(MetadataEntity metadataEntity, String metadataKey, String index, String indexColumn) {
+    MDSKey mdsIndexKey = MetadataKey.getMDSIndexKey(metadataEntity, metadataKey, index.toLowerCase());
+    String namespacedIndex = metadataEntity.getValue(MetadataEntity.NAMESPACE) + KEYVALUE_SEPARATOR +
```

Comment:
is there always a namespace? Or could a metadata entity not have a namespace? For example, a principal is at the level of the CDAP instance and has no namespace. What about custom entities?

Reply:
Also I am not sure I understand what this is for? What kind of search is satisfied with this?

Reply:
ok